specification/dartLangSpec.tex

```diff
@@ -11275,35 +11275,56 @@ \subsection{Function Expressions}
 where inferred types have already been added.%
 }
 
+\LMHash{}%
+We define the
+\IndexCustom{future type of a type}{type!future type of}
+$T$ as follows:
+
+\begin{itemize}
+\item If $T$ implements \code{Future<$S$>}
+  (\ref{interfaceSuperinterfaces})
+  then the future type of $T$ is \code{Future<$S$>}.
+\item Otherwise, if $T$ is \code{FutureOr<$S$>} bounded
+  (\ref{bindingActualsToFormals})
+  then the future type of $T$ is \code{FutureOr<$S$>}.
+\item Otherwise, if $T$ is \code{Future<$S$>?} bounded
+  or \code{FutureOr<$S$>?} bounded
+  then the future type of $T$ is \code{Future<$S$>?}
+  respectively \code{FutureOr<$S$>?}.
+\item Otherwise, $T$ has no future type.
+\end{itemize}
+
 \LMHash{}%
 We define the auxiliary function
 \IndexCustom{\flatten{T}}{flatten(t)@\emph{flatten}$(T)$},
 which is used below and in other sections, as follows:
 
 \begin{itemize}
-\item{} If $T$ is \code{$S$?}\ bounded
-  (\ref{bindingActualsToFormals})
-  for some $S$ then \DefEquals{\flatten{T}}{\code{\flatten{S}?}}.
+\item If $T$ is \code{$S$?}\ for some $S$
+  then \DefEquals{\flatten{T}}{\code{\flatten{S}?}}.
 
-\item{} If $T$ is \code{FutureOr<$S$>} bounded
-  then \DefEquals{\flatten{T}}{S}.
+\item Otherwise, if $T$ is \code{$X$\,\,\&\,\,$S$}
+  for some type variable $X$ and type $S$ then
+  \begin{itemize}
+  \item if $S$ has future type $U$
+    then \DefEquals{\flatten{T}}{\code{\flatten{U}}}.
+  \item otherwise,
+    \DefEquals{\flatten{T}}{\code{\flatten{X}}}.
+  \end{itemize}
 
-\item{} If $T$ implements \code{Future<$S$>}
-  (\ref{interfaceSuperinterfaces})
+\item Otherwise, if $T$ has future type \code{Future<$S$>}
+  or \code{FutureOr<$S$>}
   then \DefEquals{\flatten{T}}{S}.
 
-\item{} Otherwise, \DefEquals{\flatten{T}}{T}.
+\item Otherwise, if $T$ has future type \code{Future<$S$>?}\ or
+  \code{FutureOr<$S$>?}\ then \DefEquals{\flatten{T}}{\code{$S$?}}.
```

Comment:
Can these two be combined into;

* If `T` implements `Future<S>` for some `S`, _flatten_(`T`) = `S`.
* Otherwise, if `T` has future type `F`, then _flatten_(`T`) = _flatten_(`F`).

and then we rely on all the possible kinds of future types being base-cases for _flatten_?

But that's not much shorter, because what you have here is basically (if we don't use `FutureOr` for future types):

* If `T` has future type `F`:
    * if `F` is `Future<S>` for some `S`, _flatten_(`T`) = `S`
    * if `F` is `Future<S>?` for some `S`, _flatten_(`T`) = `S?`

and that's short too.

Either should be correct, and give the same result.

Reply:
The kind of simplicity I've strived for is the following: These rules specify a particular type which is being searched in a nominal-supertype hierarchy ('implements' in general, 'bounded' in the case where we can't have any interfacetype-superinterface edges in the graph), and then it directly gives the result of the search.

The simplification that I'd very much like to have is to say "if `T` has nominal-supertype `T0` then the future type is `T0`" without duplicating `T0`.

Reply:
Did that!